pkg/agent/openflow/packetin.go

```diff
@@ -15,63 +15,101 @@
 package openflow
 
 import (
+	"errors"
+	"fmt"
+
 	"github.com/contiv/ofnet/ofctrl"
 	"k8s.io/client-go/util/workqueue"
 	"k8s.io/klog"
 )
 
-type ofpPacketInReason uint
+type ofpPacketInReason uint8
 
 type PacketInHandler interface {
 	HandlePacketIn(pktIn *ofctrl.PacketIn) error
 }
 
 const (
-	// Action explicitly output to controller.
-	ofprAction ofpPacketInReason = 1
 	// Max packetInQueue size.
 	packetInQueueSize int = 256
+	// PacketIn reasons
+	PacketInReasonTF ofpPacketInReason = 1
+	PacketInReasonNP ofpPacketInReason = 0
 )
 
-func (c *client) RegisterPacketInHandler(packetHandlerName string, packetInHandler interface{}) {
+func (c *client) RegisterPacketInHandler(packetHandlerReason uint8, packetHandlerName string, packetInHandler interface{}) {
 	handler, ok := packetInHandler.(PacketInHandler)
 	if !ok {
-		klog.Errorf("Invalid Traceflow controller.")
+		klog.Errorf("Invalid controller to handle packetin.")
 		return
 	}
-	c.packetInHandlers[packetHandlerName] = handler
+	if c.packetInHandlers[packetHandlerReason] == nil {
+		c.packetInHandlers[packetHandlerReason] = map[string]PacketInHandler{}
+	}
+	c.packetInHandlers[packetHandlerReason][packetHandlerName] = handler
 }
 
-func (c *client) StartPacketInHandler(stopCh <-chan struct{}) {
-	if len(c.packetInHandlers) == 0 {
-		return
-	}
-	ch := make(chan *ofctrl.PacketIn)
-	err := c.SubscribePacketIn(uint8(ofprAction), ch)
-	if err != nil {
-		klog.Errorf("Subscribe PacketIn failed %+v", err)
-		return
-	}
-	packetInQueue := workqueue.NewNamed("packetIn")
-	go c.parsePacketIn(packetInQueue, stopCh)
+type FeatureStartPacketIn interface {
+	ConfigPacketIn()
+	ListenPacketIn()
+}
+
+type featureStartPacketIn struct {
+	reason        uint8
+	subscribeCh   chan *ofctrl.PacketIn
+	stopCh        <-chan struct{}
+	packetInQueue *workqueue.Type
+}
+
+func (f *featureStartPacketIn) ConfigPacketIn() {
+	f.subscribeCh = make(chan *ofctrl.PacketIn)
+	f.packetInQueue = workqueue.NewNamed(string(f.reason))
```

Comment:
A `newFeatureStartPacketIn ` function seems more proper than letting client construct the struct then call a method to initialize it.

Reply:
Yeah that makes sense, changing it.